db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/HashableJoinStrategy.java

```diff
@@ -133,19 +138,60 @@ public boolean feasible(Optimizable innerTable,
         }
         */
         if (innerTable.isBaseTable()) {
-			/* Must have an equijoin on a column in the conglomerate */
+            /* Must have an equijoin on a column in the conglomerate */
             cd = innerTable.getCurrentAccessPath().getConglomerateDescriptor();
         }
 
-		/* Look for equijoins in the predicate list */
-        hashKeyColumns = findHashKeyColumns(innerTable,cd,predList);
+        /* Look for equijoins in the predicate list */
+        hashKeyColumns = findHashKeyColumns(innerTable, cd, predList);
 
         if (SanityManager.DEBUG) {
             if (hashKeyColumns == null) {
-                tracer.trace(OptimizerFlag.HJ_SKIP_NO_JOIN_COLUMNS,0,0,0.0,null);
+                if (skipKeyCheck)
+                    tracer.trace(OptimizerFlag.HJ_NO_EQUIJOIN_COLUMNS, 0, 0, 0.0, null);
+                else
+                    tracer.trace(OptimizerFlag.HJ_SKIP_NO_JOIN_COLUMNS, 0, 0, 0.0, null);
+            } else {
+                tracer.trace(OptimizerFlag.HJ_HASH_KEY_COLUMNS, 0, 0, 0.0, hashKeyColumns);
             }
-            else {
-                tracer.trace(OptimizerFlag.HJ_HASH_KEY_COLUMNS,0,0,0.0,hashKeyColumns);
+        }
+        // Allow inequality join if we're skipping the key check and this is
+        // truly a join (we have a predList with join terms).
+        // Also don't allow a VALUES list, eg.,
+        //       select * from (values ('a'), ('b') ... ('z')) mytab,
+        // constructed with UnionNodes and RowResultSetNodes, to be misconstrued as a join.
+        // The same conditions as in NestedLoopJoinStrategy.feasible are added for safety.
+        // Could these be removed in the future?
+        if (hashKeyColumns == null && skipKeyCheck) {
```

Comment:
The check in estimateCost() is just to qualify whether the class method getdataSetProcessorTypeForAccessPath is available.  All other restrictions on the join are placed in feasible().  The isSpark() check is placed in estimateCost() in case it is a little expensive.  Then we call it as late as possible (I think there are more conditions besides the feasible() check that may prevent estimateCost() from being called).

Reply:
Ok, thanks.